src/dune_lang/template.mli

```diff
@@ -23,10 +23,12 @@ type t = Types.Template.t =
   }
 
 val to_string : t -> syntax:File_syntax.t -> string
+val compare_no_loc : t -> t -> Ordering.t
 val string_of_var : var -> string
 
 val pp : File_syntax.t -> Format.formatter -> t -> unit
 
 val pp_split_strings : Format.formatter -> t -> unit
 
 val remove_locs : t -> t
+val is_prefix : t -> prefix:string -> bool
```

Comment:
The `false` result is slightly ambiguous here, as we don't know whether `prefix` is really not a prefix or wether we can't answer accurately  because the string is not evaluated enough. I suggest to return something else that `bool` that can express this third state. The case where we don't know should be effectively handled the same way as `false`, but at least it's explicit when reading the code at the call site. For instance:

```
type is_prefix_result = Yes | No | Unknown
```

Reply:
Or I could rename the function into "is_prefix_before_expansion" ?

Note that the implementation of is_prefix is not trivial and could be replaced by the simpler 
```ocaml
  let is_prefix {parts; _} ~prefix = function Text s :: _ -> String.is_prefix s ~prefix | [] | Var _ :: _ -> String.length prefix = 0
```
But it assumes that text parts are always interleaved with variables (as does the already existing "text_only" function). I actually think that this simpler implementation is better, what do you think ?

Reply:
`is_prefix_before_expansion` makes me think of testing `is_prefix` on the original string, i.e. `is_prefix_before_expansion t ~prefix = String.is_prefix (to_string t) ~prefix`.

> But it assumes that text parts are always interleaved with variables (as does the already existing "text_only" function). I actually think that this simpler implementation is better, what do you think ?

Agreed. That seems better to me as well. The various functions are also careful about ensuring that we never produce `Text x :: Text y` or `Text ""`.
￼

Reply:
I like the 3-way `is_prefix_result` better than trying to express this in the name.

Regarding canonical representation, I think it's better to not rely on canonical representation if we can help it.

Maybe a nice middle ground could be a function like this:

```
let get_known_prefix = 
  let rec go t acc = match t with
    | Text s :: rest -> go (s :: acc) rest
    | _ -> String.concat ~sep:"" (List.rev acc)
  in
  fun t -> go t []
```

(or perhaps a variation that distinguishes fully known string from partially unknown)

In fact I need something like this (for the suffix) in #2073.

Reply:
FYI, the `is_prefix` that assume the canonical representation would be: 
```ocaml
type is_prefix_result = Yes | No | Unknown
let is_prefix t ~prefix =
  match t.template_.parts with
  | [Text s] -> if String.is_prefix s ~prefix then Yes else No
  | Text s :: _ ->
    if String.is_prefix s ~prefix then
      Yes
    else
      if String.is_prefix prefix ~prefix:s then
        Unknown
      else No
  | [] -> if String.length prefix = 0 then Yes else No
  | Var _ :: _ -> if String.length prefix = 0 then Yes else Unknown
```
I think I like the "get_known_prefix" better.